File: packages/ios/Sources/Components/RoutePreviewPanel/RoutePreviewPanel.swift
import SwiftUI

/// A route preview with selectable options and a continue action.
///
/// Mirrors the React `RoutePreviewPanel`. The panel never changes its own
/// selected route, and continuation is disabled while the route status is
/// `calculating`, `noRoute`, `error`, or `idle`.
public struct KozmosRoutePreviewPanel<StatusContent: View, AlertContent: View>: View {
    private let destinationName: String
    private let destinationLabel: String
    private let options: [KozmosRouteOptionPresentation]
    private let status: KozmosRouteReadiness
    private let backLabel: String
    private let continueLabel: String
    private let optionsLabel: String
    private let optionsCountLabel: String?
    private let selectedRouteAnnouncement: String?
    private let onOptionSelect: (String) -> Void
    private let onBack: () -> Void
    private let onContinue: (String) -> Void
    private let statusContent: StatusContent
    private let alert: AlertContent
    private let hasStatusContent: Bool
    private let hasAlert: Bool

    public init(
        destinationName: String,
        options: [KozmosRouteOptionPresentation],
        status: KozmosRouteReadiness,
        backLabel: String,
        continueLabel: String,
        destinationLabel: String = "To",
        optionsLabel: String = "Route options",
        optionsCountLabel: String? = nil,
        selectedRouteAnnouncement: String? = nil,
        onOptionSelect: @escaping (String) -> Void,
        onBack: @escaping () -> Void,
        onContinue: @escaping (String) -> Void,
        @ViewBuilder statusContent: () -> StatusContent,
        @ViewBuilder alert: () -> AlertContent
    ) {
        self.destinationName = destinationName
        self.options = options
        self.status = status
        self.backLabel = backLabel
        self.continueLabel = continueLabel
        self.destinationLabel = destinationLabel
        self.optionsLabel = optionsLabel
        self.optionsCountLabel = optionsCountLabel
        self.selectedRouteAnnouncement = selectedRouteAnnouncement
        self.onOptionSelect = onOptionSelect
        self.onBack = onBack
        self.onContinue = onContinue
        self.statusContent = statusContent()
        self.alert = alert()
        self.hasStatusContent = StatusContent.self != EmptyView.self
        self.hasAlert = AlertContent.self != EmptyView.self
    }

    private var selectedOption: KozmosRouteOptionPresentation? {
        options.first { $0.selected && $0.available }
    }

    private var ready: Bool { status == .ready }

    public var body: some View {
        VStack(spacing: 0) {
            VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing50) {
                Text(destinationLabel.uppercased())
                    .font(.caption.weight(.semibold))
                    .foregroundColor(KozmosColors.primitivesColorsForeground500)

                Text(destinationName)
                    .font(.title3.weight(.semibold))
                    .foregroundColor(KozmosColors.primitivesColorsForeground100)
                    .lineLimit(1)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(KozmosDimensions.primitivesLayoutSpacing200)

            Divider().overlay(KozmosColors.primitivesColorsForeground300)

            ScrollView {
                VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                    if let selectedRouteAnnouncement {
                        Color.clear
                            .frame(width: 0, height: 0)
                            .accessibilityElement()
                            .accessibilityLabel(selectedRouteAnnouncement)
                            .accessibilityAddTraits(.updatesFrequently)
                    }

                    if status != .ready, hasStatusContent {
                        statusContent
                            .frame(maxWidth: .infinity)
                            .padding(KozmosDimensions.primitivesLayoutSpacing300)
                            .background(KozmosColors.primitivesColorsBackground100.opacity(0.4))
                            .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous))
                            .overlay(
                                RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous)
                                    .strokeBorder(
                                        KozmosColors.primitivesColorsForeground300,
                                        style: StrokeStyle(lineWidth: 1, dash: [4, 4])
                                    )
                            )
                    } else {
                        ScrollView(.horizontal, showsIndicators: false) {
                            HStack(alignment: .top, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                                ForEach(options) { option in
                                    KozmosRouteOptionCard(option: option, onSelect: onOptionSelect)
                                        .frame(width: 208)
                                }
                            }
                        }
                        .accessibilityElement(children: .contain)
                        .accessibilityLabel(optionsLabel)

                        if options.count > 1, let optionsCountLabel {
                            Text(optionsCountLabel)
                                .font(KozmosTypography.caption)
                                .foregroundColor(KozmosColors.primitivesColorsForeground500)
                        }
                    }

                    if hasAlert {
                        alert
                            .font(KozmosTypography.subheadline)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(KozmosDimensions.primitivesLayoutSpacing150)
                            .background(KozmosColors.primitivesColorsEmotionalAlert500.opacity(0.15))
                            .clipShape(
                                RoundedRectangle(
                                    cornerRadius: KozmosDimensions.semanticsRadiusControl,
                                    style: .continuous
                                )
                            )
                    }
                }
                .padding(KozmosDimensions.primitivesLayoutSpacing200)
            }

            Divider().overlay(KozmosColors.primitivesColorsForeground300)

            HStack(spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                KozmosIconButton(iconName: "arrow.left", variant: .outline, action: onBack)
                    .accessibilityLabel(backLabel)

                KozmosButton(
                    continueLabel,
                    isDisabled: !ready || selectedOption == nil,
                    isLoading: status == .calculating
                ) {
                    if let selectedOption { onContinue(selectedOption.id) }
                }
                .frame(maxWidth: .infinity)
            }
            .padding(KozmosDimensions.primitivesLayoutSpacing200)
        }
        .frame(maxWidth: .infinity)
        .background(KozmosColors.primitivesColorsBackground0)
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Route preview")
    }
}

public extension KozmosRoutePreviewPanel where AlertContent == EmptyView {
    init(
        destinationName: String,
        options: [KozmosRouteOptionPresentation],
        status: KozmosRouteReadiness,
        backLabel: String,
        continueLabel: String,
        destinationLabel: String = "To",
        optionsLabel: String = "Route options",
        optionsCountLabel: String? = nil,
        selectedRouteAnnouncement: String? = nil,
        onOptionSelect: @escaping (String) -> Void,
        onBack: @escaping () -> Void,
        onContinue: @escaping (String) -> Void,
        @ViewBuilder statusContent: () -> StatusContent
    ) {
        self.init(
            destinationName: destinationName,
            options: options,
            status: status,
            backLabel: backLabel,
            continueLabel: continueLabel,
            destinationLabel: destinationLabel,
            optionsLabel: optionsLabel,
            optionsCountLabel: optionsCountLabel,
            selectedRouteAnnouncement: selectedRouteAnnouncement,
            onOptionSelect: onOptionSelect,
            onBack: onBack,
            onContinue: onContinue,
            statusContent: statusContent,
            alert: { EmptyView() }
        )
    }
}

public extension KozmosRoutePreviewPanel where StatusContent == EmptyView, AlertContent == EmptyView {
    init(
        destinationName: String,
        options: [KozmosRouteOptionPresentation],
        status: KozmosRouteReadiness,
        backLabel: String,
        continueLabel: String,
        destinationLabel: String = "To",
        optionsLabel: String = "Route options",
        optionsCountLabel: String? = nil,
        selectedRouteAnnouncement: String? = nil,
        onOptionSelect: @escaping (String) -> Void,
        onBack: @escaping () -> Void,
        onContinue: @escaping (String) -> Void
    ) {
        self.init(
            destinationName: destinationName,
            options: options,
            status: status,
            backLabel: backLabel,
            continueLabel: continueLabel,
            destinationLabel: destinationLabel,
            optionsLabel: optionsLabel,
            optionsCountLabel: optionsCountLabel,
            selectedRouteAnnouncement: selectedRouteAnnouncement,
            onOptionSelect: onOptionSelect,
            onBack: onBack,
            onContinue: onContinue,
            statusContent: { EmptyView() },
            alert: { EmptyView() }
        )
    }
}

File: packages/android/src/main/java/com/kozmos/components/RoutePreviewPanel/RoutePreviewPanel.kt
package com.kozmos.components.routepreviewpanel

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Divider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.kozmos.components.button.KozmosButton
import com.kozmos.components.iconbutton.KozmosIconButton
import com.kozmos.components.iconbutton.KozmosIconButtonVariant
import com.kozmos.components.routeoptioncard.KozmosRouteOptionCard
import com.kozmos.contracts.KozmosRouteOptionPresentation
import com.kozmos.contracts.KozmosRouteReadiness
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions

/**
 * A route preview with selectable options and a continue action.
 *
 * Mirrors the React `RoutePreviewPanel`. The panel never changes its own
 * selected route, and continuation is disabled while the route status is
 * `Calculating`, `NoRoute`, `Error`, or `Idle`.
 */
@Composable
fun KozmosRoutePreviewPanel(
    destinationName: String,
    options: List<KozmosRouteOptionPresentation>,
    status: KozmosRouteReadiness,
    backLabel: String,
    continueLabel: String,
    onOptionSelect: (String) -> Unit,
    onBack: () -> Unit,
    onContinue: (String) -> Unit,
    modifier: Modifier = Modifier,
    destinationLabel: String = "To",
    optionsLabel: String = "Route options",
    optionsCountLabel: String? = null,
    selectedRouteAnnouncement: String? = null,
    statusContent: (@Composable () -> Unit)? = null,
    alert: (@Composable () -> Unit)? = null
) {
    val selectedOption = options.firstOrNull { it.selected && it.available }
    val ready = status == KozmosRouteReadiness.Ready

    // The options list scrolls between a pinned header and a pinned footer,
    // which requires a bounded height. When the caller nests the panel
    // somewhere unbounded (another scroll container, wrapContentSize) Compose
    // would throw, so fall back to growing naturally — the same way the web
    // panel behaves.
    BoxWithConstraints(modifier = modifier.fillMaxWidth()) {
    val bodyScrolls = constraints.hasBoundedHeight
    val bodyScrollState = rememberScrollState()

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(KozmosColors.primitivesColorsBackground0)
            .semantics { contentDescription = "Route preview" }
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(KozmosDimensions.primitivesLayoutSpacing200),
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing50)
        ) {
            Text(
                text = destinationLabel.uppercase(),
                style = MaterialTheme.typography.labelSmall,
                fontWeight = FontWeight.SemiBold,
                color = KozmosColors.primitivesColorsForeground500
            )
            Text(
                text = destinationName,
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.SemiBold,
                color = KozmosColors.primitivesColorsForeground100,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
        }

        Divider(color = KozmosColors.primitivesColorsForeground300)

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .then(
                    if (bodyScrolls) {
                        Modifier
                            .weight(1f, fill = false)
                            .verticalScroll(bodyScrollState)
                    } else {
                        Modifier
                    }
                )
                .padding(KozmosDimensions.primitivesLayoutSpacing200),
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing150)
        ) {
            if (selectedRouteAnnouncement != null) {
                Box(
                    modifier = Modifier
                        .size(1.dp)
                        .semantics {
                            contentDescription = selectedRouteAnnouncement
                            liveRegion = LiveRegionMode.Polite
                        }
                )
            }

            if (status != KozmosRouteReadiness.Ready && statusContent != null) {
                Surface(
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel),
                    color = KozmosColors.primitivesColorsBackground100.copy(alpha = 0.4f),
                    border = BorderStroke(1.dp, KozmosColors.primitivesColorsForeground300)
                ) {
                    Box(
                        modifier = Modifier.padding(KozmosDimensions.primitivesLayoutSpacing300),
                        contentAlignment = Alignment.Center
                    ) {
                        statusContent()
                    }
                }
            } else {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .horizontalScroll(rememberScrollState())
                        .semantics { contentDescription = optionsLabel },
                    horizontalArrangement = Arrangement.spacedBy(
                        KozmosDimensions.primitivesLayoutSpacing150
                    )
                ) {
                    options.forEach { option ->
                        KozmosRouteOptionCard(
                            option = option,
                            onSelect = onOptionSelect,
                            modifier = Modifier.width(208.dp)
                        )
                    }
                }

                if (options.size > 1 && optionsCountLabel != null) {
                    Text(
                        text = optionsCountLabel,
                        style = MaterialTheme.typography.bodySmall,
                        color = KozmosColors.primitivesColorsForeground500
                    )
                }
            }

            if (alert != null) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
                        .background(KozmosColors.primitivesColorsEmotionalAlert500.copy(alpha = 0.15f))
                        .padding(KozmosDimensions.primitivesLayoutSpacing150)
                ) {
                    alert()
                }
            }
        }

        Divider(color = KozmosColors.primitivesColorsForeground300)

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(KozmosDimensions.primitivesLayoutSpacing200),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing150)
        ) {
            KozmosIconButton(
                icon = Icons.AutoMirrored.Filled.ArrowBack,
                onClick = onBack,
                contentDescription = backLabel,
                variant = KozmosIconButtonVariant.Outline
            )

            KozmosButton(
                onClick = { selectedOption?.let { onContinue(it.id) } },
                modifier = Modifier.weight(1f),
                enabled = ready && selectedOption != null,
                isLoading = status == KozmosRouteReadiness.Calculating
            ) {
                Text(continueLabel)
            }
        }
    }
    }
}
